### PR TITLE
Security token tests

### PR DESCRIPTION
**Pull request by man-of-code**
Tests for Security Token
Rebase with Master

### CODE REVIEW

File: examples/permissioned-token/test/tests.js [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    runtime.optIntoASA(assetIndex, acc.address, {});
+    syncInfo();
+  }
+
+  function optInToPermissions (address) {
+    runtime.optInToApp(address, permissionsAppId, {}, {});
+    syncInfo();
+  }
+
+  function issue (acc, amount) {
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:47]: Please add docs, what does this function do ?
how should transaction be sent ?

File: examples/permissioned-token/test/tests.js [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+      fromAccount: alice.account,
+      appId: controllerAppID,
+      payFlags: { totalFee: 1000 },
+      appArgs: ['str:kill'],
+      foreignAssets: [assetIndex]
+    });
+    syncInfo();
+  }
+
+  function whitelist (address) {
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:87]: Please add docs for all the functions

File: examples/permissioned-token/test/tests.js [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    alice = new AccountStore(minBalance, { addr: ALICE_ADDRESS, sk: new Uint8Array(0) });
+    bob = new AccountStore(minBalance);
+    elon = new AccountStore(minBalance);
+    runtime = new Runtime([master, alice, bob, elon]);
+
+    // Deploy ASA
+    assetIndex = runtime.addAsset('gold', { creator: { ...alice.account, name: 'alice' } });
+    asaDef = runtime.getAssetDef(assetIndex);
+
+    // Setup Controller SSC
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:170]:
```suggestion
    // Deploy Controller SSC
```

File: examples/permissioned-token/test/tests.js [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+      globalBytes: 2,
+      appArgs: [`int:${assetIndex}`],
+      foreignAssets: [assetIndex]
+    };
+    CONTROLLER_PROGRAM = getProgram(CONTROLLER, { TOKEN_ID: assetIndex });
+    controllerAppID = runtime.addApp(
+      sscFlags, {}, CONTROLLER_PROGRAM, CLEAR_STATE_PROGRAM
+    );
+
+    // Setup Permissions SSC
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:185]:
```suggestion
    // Deploy Permissions SSC
```

File: examples/permissioned-token/test/tests.js [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    syncInfo();
+  });
+
+  it('Token Issuance', () => {
+    // Cannot issue before opting-in
+    assert.throws(() => issue(elon, 20), 'RUNTIME_ERR1404');
+
+    // Can issue after optin-in
+    optInToASA(elon);
+    // syncInfo();
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:248]: do we need this ?

File: examples/permissioned-token/test/tests.js [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    });
+    runtime.optIntoASA(assetIndex, clawbackAddress, {});
+
+    // Refresh Accounts
+    syncInfo();
+  });
+
+  it('Token Issuance', () => {
+    // Cannot issue before opting-in
+    assert.throws(() => issue(elon, 20), 'RUNTIME_ERR1404');
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:244]: Please add tests in two flows, one with passing flow(with correct calls, etc), other one is failing paths(which will have corner cases, failing cases).

File: examples/permissioned-token/test/tests.js [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+    assert.equal(
+      Number(runtime.getAssetHolding(assetIndex, elon.address).amount),
+      Number(elonBalance) + amount
+    );
+    assert.equal(
+      Number(runtime.getAssetHolding(assetIndex, elon.address).amount),
+      Number(bobBalance) + amount
+    );
+  });
+});
```
Comment (amityadav0) [inline examples/permissioned-token/test/tests.js:334]: Please add more tests and divide them in different files if needed, so it could be easily understandable for a developer

File: examples/permissioned-token/test/tests.js (hunk 6 above)
Comment (man-of-code) [reply examples/permissioned-token/test/tests.js:244]: Are you referring to 2 separate files?

File: examples/permissioned-token/test/tests.js (hunk 7 above)
Comment (man-of-code) [reply examples/permissioned-token/test/tests.js:334]: Okay

File: examples/permissioned-token/test/tests.js (hunk 5 above)
Comment (man-of-code) [reply examples/permissioned-token/test/tests.js:248]: No
